### PR TITLE
Fix exception in AS::Timezone.all when any tzinfo data is missing

### PR DESCRIPTION
**Pull request by dsander**
Before this change missing timezone data for any of the time zones defined in `ActiveSupport::Timezone::MAPPING` caused a `comparison of NilClass with ActiveSupport::TimeZone failed` exception.

Attempting to get a timezone by passing a number/duration to `[]` or calling `all` directly will try to sort sort the values of `zones_map`. Those values are initialized by the return value of `create(zonename)` which returns `nil` if `TZInfo` is unable to find the timezone information.

In our case the exception was triggered by an outdated tzdata package which did not include information for the "recently" added time zones.

Before 078421bacba178eac6a8e607b16f3f4511c5d72f `zones_map` only returned the information that have been loaded into `@lazy_zone_map` which ignored time zones for which the data could not be loaded, this change restores the previous behaviour.

### CODE REVIEW

File: activesupport/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+*   Fix bug where `ActiveSupport::Timezone.all` would fail when tzinfo data for
+    any timezone defined in `ActiveSupport::MAPPING` is missing.
```
Comment (rafaelfranca) [inline activesupport/CHANGELOG.md]: How would this happen in real world? All zones defined in that constant we know exists.

Comment (dsander) [reply activesupport/CHANGELOG.md]: They do exist if the `tzdata` system package is current, but when it is not `TZInfo` does not know anything about them and raises an exception itself. Before 078421bacba178eac6a8e607b16f3f4511c5d72f `AS::Timezone.all` just omitted those timezones.

File: activesupport/test/time_zone_test_helpers.rb [hunk 2]
```diff
@@ -23,4 +23,15 @@ def with_preserve_timezone(value)
...
+  def with_missing_tzdata_for(timezone_name)
+    klass = Class.new(ActiveSupport::TimeZone) do
+      define_method :initialize do |name, utc_offset = nil, tzinfo = nil|
+        raise TZInfo::InvalidTimezoneIdentifier if name == timezone_name
+        super(name, utc_offset, tzinfo)
+      end
+    end
+    klass.clear
+    yield klass
+  end
```
Comment (pixeltrix) [inline activesupport/test/time_zone_test_helpers.rb]: Replace this method with this:

``` ruby
def with_tz_mappings(mappings)
  old_mappings = ActiveSupport::TimeZone::MAPPING.dup
  ActiveSupport::TimeZone.clear
  ActiveSupport::TimeZone::MAPPING.clear
  ActiveSupport::TimeZone::MAPPING.merge!(mappings)

  yield
ensure
  ActiveSupport::TimeZone.clear
  ActiveSupport::TimeZone::MAPPING.clear
  ActiveSupport::TimeZone::MAPPING.merge!(old_mappings)
end
```

and use as explained above.

File: activesupport/test/time_zone_test.rb [hunk 3]
```diff
@@ -725,6 +725,15 @@ def test_all_uninfluenced_by_time_zone_lookups_delegated_to_tzinfo
...
 
+  def test_all_not_raises_exception_with_mizzing_tzinfo_data
+    with_missing_tzdata_for("Puerto Rico") do |timezone|
+      assert_nil timezone["Puerto Rico"]
+      assert_nothing_raised do
+        timezone.all
+      end
+    end
+  end
+
```
Comment (pixeltrix) [inline activesupport/test/time_zone_test.rb]: While creating a subclass works as a way of reproducing the error it doesn't really reflect real-life usage so I'd prefer a test that more accurately reflects the error which we can do by changing the mappings for this test using the helper method in the comment below. With that you can then change the test to this:

``` ruby
def test_all_not_raises_exception_with_mizzing_tzinfo_data
  mappings = {
    "Puerto Rico" => "America/Unknown",
    "Pittsburgh"  => "America/New_York"
  }

  with_tz_mappings(mappings) do
    assert_nil ActiveSupport::TimeZone["Puerto Rico"]
    assert_nil ActiveSupport::TimeZone[-9]
    assert_nothing_raised do
      ActiveSupport::TimeZone.all
    end
  end
end
```

**Review by pixeltrix — CHANGES_REQUESTED** [review]
I think the change is 👍  we just need to change the tests as I've outlined, thanks.

File: activesupport/test/time_zone_test.rb (hunk 3 above)
Comment (dsander) [reply activesupport/test/time_zone_test.rb]: Thanks, I wasn't particular happy with creating the subclass and now realize that I created the test from the wrong "perspective".

**Comment by dsander** [comment]
@pixeltrix Thanks for the feedback, should I squash the commits into one?

**Comment by rafaelfranca** [comment]
> should I squash the commits into one?

Yes

**Review by pixeltrix — APPROVED** [review]

**Comment by pixeltrix** [comment]
@dsander thanks! 👍

**Pull request merged**